core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,403 @@
+                values_state_answer.append(value)
+
+        state_stats_mapping = {}
+        # Sort list of state hit events in increasing order of it's timestamp.
+        values_state_hit = sorted(values_state_hit, key=lambda k: k[
+            'created_on'])
+        for version in range(1, latest_exp_version+1):
+            values_state_hit_versioned = [
+                val for val in values_state_hit if val['version'] == version]
+            values_state_answer_versioned = [
+                val for val in values_state_answer if val['version'] == version]
+            exploration_versioned = exp_services.get_exploration_by_id(
+                exp_id, version)
+
+            if version == 1:
+                # Create default state stats mapping for the first version.
+                for state_name in exploration_versioned.states:
+                    state_stats_mapping[state_name] = (
+                        stats_domain.StateStats.create_default())
+            else:
+                exp_commit_log = exp_models.ExplorationCommitLogEntryModel.get(
+                    'exploration-%s-%s' % (exp_id, version))
+                change_list = exp_commit_log.commit_cmds
+
+                # Handling state additions, renames and deletions.
+                for change_dict in change_list:
+                    if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+                        state_stats_mapping[change_dict['state_name']] = (
+                            stats_domain.StateStats.create_default())
+                    elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
+                        state_stats_mapping.pop(change_dict['state_name'])
+                    elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
+                        state_stats_mapping[change_dict['new_state_name']] = (
+                            state_stats_mapping.pop(
+                                change_dict['old_state_name']))
+
+            # Dict mapping each state name of the exploration and the list of
+            # unique session IDs that have entered that state.
+            state_session_ids = {
+                state_name: [] for state_name in exploration_versioned.states
+            }
+            session_ids_set = set()
+            # Compute total_hit_count and first_hit_count for the states.
+            for value in values_state_hit_versioned:
+                state_name = value['state_name']
+                state_stats_mapping[state_name].total_hit_count_v1 += 1
+                if value['session_id'] not in state_session_ids[state_name]:
+                    state_stats_mapping[state_name].first_hit_count_v1 += 1
+                    state_session_ids[state_name].append(value['session_id'])
+                    session_ids_set.add(value['session_id'])
+
+            # Compute num_completions for the states.
+            for session_id in session_ids_set:
+                values_state_hit_sessioned = [
```

Comment:
This iteration through values_state_hit_versioned multiple times for each session ID is very inefficient and will slow things down a lot. It would be better to figure out a way to do just a single pass through the list and obtain the data keyed by session ID after that single pass.